### PR TITLE
tuxedo-rs: 0.2.3 -> 0.3.0

### PR DESCRIPTION
**Pull request by mrcjkb**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mrcjkb** [comment]
Result of `nixpkgs-review pr 268559` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>tailor-gui</li>
    <li>tuxedo-rs</li>
  </ul>
</details>

**Comment by xaverdh** [comment]
builds and works on my system

File: pkgs/os-specific/linux/tuxedo-rs/default.nix
Comment (xaverdh) [inline pkgs/os-specific/linux/tuxedo-rs/default.nix:9]: Not sure if deduping `pname` is a good idea, since the repository location might one day change without the package name changing..
Deduping `version` is a good idea, I think.

Comment (mrcjkb) [reply pkgs/os-specific/linux/tuxedo-rs/default.nix:9]: I don't thinks that matters too much, does it?
I mean, if the repo location changes, we have to update the `src` anyway. I'd lean towards YAGNI on this.

Comment (xaverdh) [reply pkgs/os-specific/linux/tuxedo-rs/default.nix:9]: well either way its not too important

**Review by xaverdh — APPROVED** [review]

**Comment by xaverdh** [comment]
cc @blitz @gonzaloetjo might be interested in testing this

**Review by blitz — APPROVED** [review]

**Comment by mrcjkb** [comment]
@blitz shall I add you to the maintainers list so you get notified on future PRs?

**Comment by blitz** [comment]
> @blitz shall I add you to the maintainers list so you get notified on future PRs?

Please don't. I'm planning to retire my tuxedo soon.

**Comment by mrcjkb** [comment]
Result of `nixpkgs-review pr 268559` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>tailor-gui</li>
    <li>tuxedo-rs</li>
  </ul>
</details>

**Comment by xaverdh** [comment]
> @blitz @xaverdh shall I add you to the maintainers list so you get notified on future PRs?

you may add me

**Review by xaverdh — APPROVED** [review]
Thanks for adding me!

**Comment by xaverdh** [comment]
this has been sitting here for quite long.. cc @SuperSandro2000

**Review by blitz — APPROVED** [review]
Maybe @nikstur can help?

File: pkgs/os-specific/linux/tuxedo-rs/default.nix
Comment (nikstur) [inline pkgs/os-specific/linux/tuxedo-rs/default.nix:24]: It is generally undesired to use variables (besides version) to populate fetchFromGitHub because it makes overriding harder.

File: pkgs/os-specific/linux/tuxedo-rs/default.nix
Comment (nikstur) [inline pkgs/os-specific/linux/tuxedo-rs/default.nix:55]: The way this was done previously is better, also because this doesn't typically change much.

File: pkgs/os-specific/linux/tuxedo-rs/default.nix
Comment (nikstur) [inline pkgs/os-specific/linux/tuxedo-rs/default.nix:7]: Instead of using the `let ... in` construction in `callPackage` derivations it is generally better to use `rec` and re-use the variables inside `buildRustPackage` or `mkDerivation`. This allows to call `overrideAttrs` on the derivation and achieve the desired results.

File: pkgs/os-specific/linux/tuxedo-rs/default.nix
Comment (mrcjkb) [reply pkgs/os-specific/linux/tuxedo-rs/default.nix:24]: question: do you have a link to a guideline for this?

I see

```nix
    repo = pname;
    rev = "v${version}";
```

in many packages, and typically overrides just override the entire `src` attrset (as the hash changes, too).

File: pkgs/os-specific/linux/tuxedo-rs/default.nix
Comment (mrcjkb) [reply pkgs/os-specific/linux/tuxedo-rs/default.nix:7]: TIL... 🙏

**Comment by mrcjkb** [comment]
Result of `nixpkgs-review pr 268559` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>tailor-gui</li>
    <li>tuxedo-rs</li>
  </ul>
</details>

File: pkgs/os-specific/linux/tuxedo-rs/default.nix
Comment (nikstur) [reply pkgs/os-specific/linux/tuxedo-rs/default.nix]: Sadly, I don't have a guideline for this but I think I can explain why this is needed: If you call `overrideAttrs` on such a derivation trying to change `pname`, the fetcher would not be able to retrieve the sources anymore. However, sometimes you want to change the `pname` without changing the sources. For example, we do this for systemdMinimal.

Comment (mrcjkb) [reply pkgs/os-specific/linux/tuxedo-rs/default.nix]: I see, thanks. Good to know. I suppose this isn't an issue for the version? In such a case, you would probably want the source to change with the version, right?

Comment (nikstur) [reply pkgs/os-specific/linux/tuxedo-rs/default.nix]: Yes exactly!

File: pkgs/os-specific/linux/tailor-gui/default.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/tailor-gui/default.nix:20]: inheriting source and then overriding version sounds fishy to me. We should either take both or fetch the src here.

File: pkgs/os-specific/linux/tuxedo-rs/default.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/tuxedo-rs/default.nix:8]:
```suggestion
```

File: pkgs/os-specific/linux/tailor-gui/default.nix
Comment (mrcjkb) [reply pkgs/os-specific/linux/tailor-gui/default.nix:20]: Do you see any actual pitfalls to this?

The crates in tuxedo-rs have versions that update independently, but the packages still belong together, and ideally the sources shouldn't be out of sync.

I would read the versions from the respective Cargo.toml, but that would be IFD...

**Comment by mrcjkb** [comment]
Update: Ran `nixfmt`.

File: pkgs/os-specific/linux/tailor-gui/default.nix
Comment (SuperSandro2000) [reply pkgs/os-specific/linux/tailor-gui/default.nix]: tuxedo-rs could be updated and also include an update to tuxedo-gui but we could forget to update the version number here.

File: pkgs/os-specific/linux/tuxedo-rs/default.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/tuxedo-rs/default.nix:33]: I don't think this is an improvement to the code and a bad formatting decission by nixfmt.

File: pkgs/os-specific/linux/tailor-gui/default.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/tailor-gui/default.nix:29]: same here

File: pkgs/os-specific/linux/tailor-gui/default.nix
Comment (mrcjkb) [reply pkgs/os-specific/linux/tailor-gui/default.nix]: And if we use separate sources, tuxedo-rs could be updated and we'd forget to update tailor-gui, which would potentially break tailor-gui.
With shared sources, tailor-gui will fail to build if we update the source and the cargo hash changes.

This could potentially be solved with a `passthru.tests.version` for tailor-gui.
But as it stands, rust GTK apps don't support passing in a `--version` flag (afaik).
I've raised [an issue about this upstream](https://github.com/AaronErhardt/tuxedo-rs/issues/61). Maybe there's a way to do it.

Comment (nikstur) [reply pkgs/os-specific/linux/tailor-gui/default.nix]: I think this is good enough for now. No reason to block the PR over this.

**Comment by nikstur** [comment]
Result of `nixpkgs-review pr 268559` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>tailor-gui</li>
    <li>tuxedo-rs</li>
  </ul>
</details>

**Comment by nikstur** [comment]
The formatting seems like a *decrease* in quality of the package and is not something that I think should generally be done to packages that you update or otherwise change. However, I won't block on this.

**Review by nikstur — APPROVED** [review]

**Comment by mrcjkb** [comment]
> The formatting seems like a _decrease_ in quality of the package and is not something that I think should generally be done to packages that you update or otherwise change. However, I won't block on this.

I feel the same way. I'm generally not a fan of nixfmt. But often, when I fixup a commit, I accidentally leave behind an extra whitespace line (or similar) that leads to further delays. That's why I'd prefer to stick to whatever formatting tool is recommended here, even if it decreases the quality.

**Comment by nikstur** [comment]
> I'd prefer to stick to whatever formatting tool is recommended here

As of now there is NO officially recommended formatter. The closest thing is the (not yet passed) [RFC 166](https://github.com/NixOS/rfcs/pull/166). Please, note, however, that even if they do use nixmft, they use a special branch with a format that deviates (significantly) from the current nixfmt default.

**Comment by mrcjkb** [comment]
> > I'd prefer to stick to whatever formatting tool is recommended here
> 
> As of now there is NO officially recommended formatter. The closest thing is the (not yet passed) [RFC 166](https://github.com/NixOS/rfcs/pull/166). Please, note, however, that even if they do use nixmft, they use a special branch with a format that deviates (significantly) from the current nixfmt default.

Alright. I've reverted the formatting for now.

File: pkgs/os-specific/linux/tailor-gui/default.nix
Comment (SuperSandro2000) [reply pkgs/os-specific/linux/tailor-gui/default.nix:20]: Just don't forget to update this when r-ryantm comes by.

**Pull request merged**
